### PR TITLE
Update rustfmt command line interface

### PR DESCRIPTION
**Pull request by topecongiro**
This PR updates the rustfmt command-line interface in the following ways. Sorry in advance about the number of changes!

- Remove `--skip-children`.
- Remove `--help=[command]` and introduce the following alternatives:
    - `--help=config` can now be invoked via `--config help`
    - `--help=file-lines` is now shown with `--help` by default
- Change the behavior of `--print-config`
    - Remove the output file argument: now always print to stdout
    - `--print-config minimal` prints the set of current configuration that differs from the default, without formatting files.
    - `--print-config default` prints the set of all configuration options with the default value
    - `--print-config current` prints the set of the current configuration, without formatting files
- Remove `--unstable-features`.
- Remove `--color`.
- `--help` now prints longer help messages
    - This change is a side-effect of replacing `getopts` with `structopt`. Please see the following comment on how it changed.
- Change the behavior of `Config::used_options`.
    - Return a set of configuration options that were explicitly set to values that are different from the default value

### CODE REVIEW

**Comment by topecongiro** [comment]
`rustfmt -h`:

```
rustfmt 2.0.0-rc.1-nightly (265697c2 2020-01-08)
Format Rust code

USAGE:
    rustfmt [FLAGS] [OPTIONS] [--] [files]...

FLAGS:
    -c, --check                   Run in 'check' mode
        --error_on_unformatted    Error if unable to get comments or string literals within max_width, or they are left
                                  with trailing whitespaces (unstable)
    -l, --files-with-diff         Prints the names of files with diff
    -h, --help                    Prints help information
    -q, --quiet                   Print no output
    -r, --recursive               Recursively format submodules
    -V, --version                 Prints version information
    -v, --verbose                 Print verbose output

OPTIONS:
        --edition <2015|2018>                       Rust compiler edition
        --config-path <config-path>                 A path to the configuration file
        --print-config <default|current|minimal>    Print configuration options
        --emit <emit>                               Specify the format of rustfmt's output
        --file-lines <file-lines>                   Limit formatting to specified ranges [default: null]
        --config <inline-config>...                 Set options from command line

ARGS:
    <files>...    
```

`rustfmt --help`:

```
rustfmt 2.0.0-rc.1-nightly (265697c2 2020-01-08)
Format Rust code

USAGE:
    rustfmt [FLAGS] [OPTIONS] [--] [files]...

FLAGS:
    -c, --check                   
            Run in 'check' mode.
            
            Exits with 0 if input is formatted correctly. Exits with 1 and prints a diff if formatting is required.
        --error_on_unformatted    
            Error if unable to get comments or string literals within max_width, or they are left with trailing
            whitespaces (unstable)
    -l, --files-with-diff         
            Prints the names of files with diff

    -h, --help                    
            Prints help information

    -q, --quiet                   
            Print no output

    -r, --recursive               
            Recursively format submodules.
            
            Format all encountered modules recursively regardless of whether the modules are defined inline or in
            another file.
    -V, --version                 
            Prints version information

    -v, --verbose                 
            Print verbose output


OPTIONS:
        --edition <2015|2018>                       
            Rust compiler edition
            
            Specify which edition of the compiler to use when formatting code.
        --config-path <config-path>                 
            A path to the configuration file

        --print-config <default|current|minimal>    
            Print configuration options.
            
            `default` will print the default configuration options. `current` will print the current configuration
            options. `minimal` will print the minimal subset of the current configuration options.
        --emit <emit>                               
            Specify the format of rustfmt's output

        --file-lines <file-lines>                   
            Limit formatting to specified ranges.
            
            If you want to restrict reformatting to specific sets of lines, you can use the `--file-lines` option. Its
            argument is a JSON array of objects with `file` and `range` properties, where `file` is a file name, and
            `range` is an array representing a range of lines like `[7,13]`. Ranges are 1-based and inclusive of both
            end points. Specifying an empty array will result in no files being formatted. For example,
            
            ``` rustfmt --file-lines '[ {{\"file\":\"src/lib.rs\",\"range\":[7,13]}},
            {{\"file\":\"src/lib.rs\",\"range\":[21,29]}}, {{\"file\":\"src/foo.rs\",\"range\":[10,11]}},
            {{\"file\":\"src/foo.rs\",\"range\":[15,15]}}]' ```
            
            would format lines `7-13` and `21-29` of `src/lib.rs`, and lines `10-11`, and `15` of `src/foo.rs`. No other
            files would be formatted, even if they are included as out of line modules from `src/lib.rs`. [default:
            null]
        --config <inline-config>...                 
            Set options from command line.
            
            Set configuration options via command line by specifying a list of key-value pairs separated by commas
            (e.g., rustfmt --config=max_width=100,merge_imports=true). These settings precedes any other settings
            specified in configuration files.

ARGS:
    <files>...    
            
```

File: rustfmt-core/rustfmt-config/src/lib.rs [hunk 1]
```diff
@@ -169,6 +169,7 @@ impl PartialConfig {
         cloned.width_heuristics = None;
         cloned.print_misformatted_file_names = None;
         cloned.recursive = None;
+        cloned.emit_mode = None;
```
Comment (calebcartwright) [inline rustfmt-core/rustfmt-config/src/lib.rs]: I think `skip_children` up on line 130 would need to be removed too to fix the failing `test_dump_default_config` test right?

File: rustfmt-core/rustfmt-bin/src/bin/main.rs [hunk 2]
```diff
@@ -35,25 +39,247 @@ fn main() {
...
+    /// Rust compiler edition
+    ///
+    /// Specify which edition of the compiler to use when formatting code.
+    #[structopt(long, name = "2015|2018")]
+    edition: Option<Edition>,
+    /// Print configuration options.
+    ///
+    /// `default` will print the default configuration options. `current` will print the
+    /// current configuration options. `minimal` will print the minimal subset of the
+    /// current configuration options.
```
Comment (calebcartwright) [inline rustfmt-core/rustfmt-bin/src/bin/main.rs]:
```suggestion
    /// current configuration options that differ from the default configuration option values.
```

File: rustfmt-core/rustfmt-bin/src/bin/main.rs [hunk 3]
```diff
@@ -35,25 +39,247 @@ fn main() {
...
+#[structopt(name = "rustfmt", version = include_str!(concat!(env!("OUT_DIR"), "/version-info.txt")))]
+struct Opt {
+    /// Run in 'check' mode.
+    ///
+    /// Exits with 0 if input is formatted correctly.
+    /// Exits with 1 and prints a diff if formatting is required.
+    #[structopt(short, long)]
+    check: bool,
+    /// Specify the format of rustfmt's output.
+    #[structopt(long)]
```
Comment (calebcartwright) [inline rustfmt-core/rustfmt-bin/src/bin/main.rs]: Currently on stable the help text for `--emit` excludes the nightly-only emit modes:

```bash
        --emit [files|stdout]
                        What data to emit and how
```

and on nightly all the supported emit modes are listed:
```bash
        --emit [files|stdout|coverage|checkstyle|json]
                        What data to emit and how
```

Assuming we aren't stabilizing all the other emit modes as part of this PR, I think we could maintain that dynamic help behavior via:

```suggestion
    #[cfg_attr(nightly, structopt(long, name= "files|stdout|checkstyle|json"))]
    #[cfg_attr(not(nightly), structopt(long, name= "files|stdout"))]
```

File: rustfmt-core/rustfmt-bin/src/bin/main.rs [hunk 4]
```diff
@@ -35,25 +39,247 @@ fn main() {
...
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Emit {
+    Files,
+    Stdout,
+    Checkstyle,
+    Json,
+}
+
```
Comment (calebcartwright) [inline rustfmt-core/rustfmt-bin/src/bin/main.rs]: Is the removal of `Coverage` emit mode intentional?

File: rustfmt-core/rustfmt-bin/src/bin/main.rs (hunk 3 above)
Comment (topecongiro) [reply rustfmt-core/rustfmt-bin/src/bin/main.rs]: Good catch, thank you!

File: rustfmt-core/rustfmt-bin/src/bin/main.rs (hunk 4 above)
Comment (topecongiro) [reply rustfmt-core/rustfmt-bin/src/bin/main.rs]: It was not, though I am planning to remove the coverage mode in the further PRs, so I might as well leave this as is.

Comment (calebcartwright) [reply rustfmt-core/rustfmt-bin/src/bin/main.rs]: Yeah that makes sense

**Pull request merged**